tools/azure-sdk-tools/ci_tools/dependency_analysis.py

```diff
@@ -0,0 +1,350 @@
+                print("%s" % (spec if spec else "(empty)"))
+                for lib in specs[spec]:
+                    print("  * %s" % (lib))
+            print("")
+
+    ## verify dependencies
+    for package in dependencies:
+        known_reqs.append(package)
+        speclist = dependencies[package].keys()
+        dep_versions = get_known_versions(package)
+
+        if package.startswith("azure"):
+            # include local version if present
+            if package in packages:
+                dep_versions.append(Version(packages[package]["version"]))
+
+        remaining_dep_versions = list(dep_versions)
+
+        for spec in speclist:
+            if not spec:
+                # package with no specifier means that any specifier will match.
+                continue
+            spec = SpecifierSet(spec)
+
+            previous_remaining_dep_versions = list(remaining_dep_versions)
+            remaining_dep_versions = [version for version in remaining_dep_versions if str(version) in spec]
+
+            if not remaining_dep_versions:
+                exitcode = 1
+                incompatible.append(package)
+
+                if args.verbose:
+                    print(
+                        f"{package} is required in an irreconcilable combination. No available versions can be found to match the following combination of specifiers:"
+                    )
+                    for specifier in sorted(dependencies[package].keys(), key=lambda x: len(x)):
+                        if specifier:
+                            print(
+                                f"{dependencies[package][specifier]} {pluralize(dependencies[package][specifier], 'takes', 'take')} a dependency on specifier {specifier}"
+                            )
+                    print(f'Resolve conflicts in reqs for "{package}" and re-run.')
+
+    # verify requirements in shared_requirements.txt
+    if os.path.exists(frozen_filename):
+        with open(frozen_filename, "r") as f:
+            frozen_reqs = set([line.strip() for line in f.readlines()])
+
+        discovered_reqs = set(dependencies.keys())
+        # packages that are part of the dependencies and not yet in the shared_requirement.txt
+        new_reqs = discovered_reqs.difference(frozen_reqs)
+        # packages that are in the frozen list and no longer part of the discovered dependencies
+        missing_reqs = frozen_reqs.difference(discovered_reqs)
+
+        if new_reqs:
+            print(
+                f"Unknown {pluralize(new_reqs, 'dependency', 'dependencies')} found that are not present in shared_requirements.txt. Use --verbose for details."
+            )
+            if args.verbose:
+                print(f"Unknown {pluralize(new_reqs, 'dependency', 'dependencies')}: {list(new_reqs)}")
+            exitcode = 1
+    else:
+        new_reqs = set(dependencies.keys())
+        print("No shared_requirements.txt file can be located to compare against existing.")
```

Comment:
should this be an `exitcode = 1` scenario?

Reply:
You think so? I think it's fair that you can run this thing with no shared_requirements.txt.

It checks

- That no external dependencies were untracked
- That all the dependencies have at least one package that satisfies all their specifications.

I didn't necessarily think that should be a failure when you don't have a lockfile specified. I'll just make it an exitcode 1 scenario.

Reply:
I think I'm not clear on the case where we would not already have the shared_requirements.txt in the python repo. But if the idea is to make this script generic/usable in other places/repos then that makes sense to me.

Reply:
> I think I'm not clear on the case where we would not already have the shared_requirements.txt in the python repo

We wouldn't, but I'm trying to make this approachable. You question below about running with `--freeze` is all you'd need to do to _generate_ a shared_requirements. So you have a way to start one. Otherwise we just won't validate against it.

Not a code path the sdk-for-python will ever see most likely.